sdk/src/message.rs

```diff
@@ -148,13 +180,41 @@ impl Message {
         )
     }
 
-    pub fn program_ids(&self) -> &[Pubkey] {
-        &self.account_keys
-            [self.account_keys.len() - self.header.num_credit_only_unsigned_accounts as usize..]
+    pub fn program_ids(&self) -> Vec<&Pubkey> {
+        self.instructions
+            .iter()
+            .map(|ix| &self.account_keys[ix.program_ids_index as usize])
+            .collect()
+    }
+
+    pub fn program_index_in_program_ids(&self, index: usize) -> Option<usize> {
+        let program_ids = self.program_ids();
+        program_ids
```

Comment:
`position()`?

Reply:
`program_position()`?

Reply:
No, I'm wondering if you just implemented the builtin function `position()`.

Reply:
https://github.com/solana-labs/solana/blob/d77f479c762f3fcd51d64a4a6e178c7d8514df50/sdk/src/message.rs#L9

Reply:
Haha, yep! Don't know how I missed that one.

Reply:
Oh, that `position()`. I thought you meant `Iterator::position`, which I *definitely* should have used.
While I may end up reworking message_processor, at present I still need a method that converts a program's index in the list of all accounts to its index in the filtered list of program ids. I could call `position` inside such a method, except `position` panics blindly when an index isn't found, instead of returning something that can be converted into a TransactionError (https://github.com/solana-labs/solana/blob/d77f479c762f3fcd51d64a4a6e178c7d8514df50/runtime/src/message_processor.rs#L222)
Should I rework position to return an Option?

Reply:
No, thanks. My helper assumes None can never happen. I'd guess that's true in your case as well?

Reply:
That may be true. I will take a closer look at that when I address the `TODO` in this block for CO-account handling